app/src/org/commcare/network/RequestStats.kt

```diff
@@ -0,0 +1,73 @@
+package org.commcare.network
+
+import org.commcare.CommCareApp
+import org.commcare.CommCareApplication
+import org.commcare.resources.model.InstallRequestSource
+import org.commcare.utils.SyncDetailCalculations.getDaysBetweenJavaDatetimes
+import org.commcare.utils.TimeProvider
+import java.util.*
+
+/**
+ *  Used to keep track of metrics like time since last suceess for various resource install requests
+ *  Any resource install workflows from [InstallRequestSource] should [register] before starting and
+ *  [markSuccess] on a successful run
+ */
+object RequestStats {
+
+    @JvmStatic
+    fun register(installRequestSource: InstallRequestSource) {
+        register(CommCareApplication.instance().currentApp, installRequestSource)
+    }
+
+    @JvmStatic
+    fun register(app: CommCareApp, installRequestSource: InstallRequestSource) {
```

Comment:
This is going to make the SharedPreferences object enormous, right? With potentially thousands of entries?

I don't think that's a best practice for how we use these buckets. It could affect their perforamnce for what we expected to be very fast I/O lookups, and it will be hard for us to do things like migrate them. I think it would be better to store this data in its own db tables like we do for a few other models like the geocoder entity cache

Reply:
> This is going to make the SharedPreferences object enormous, right? With potentially thousands of entries?

Not really, we are not making an entry here per resource but per install workflow which gets erased on a successful run.  Furthermore, for multiple failed runs of a install workflow, we only ever maintain the time of the first failed attempt to get the time since first failed attempt. So there can only be at max 6 entries each corresponding to each of the install workflow from `InstallRequestSource`

Reply:
We had a big problem with the other update status collector object blowing up like crazy, so I still think that if we are going to keep some number of these records, we should make sure that number is a fixed LIFO queue (Even if it's fixed at some reasonably large size like 100), rather than letting it grow indefinitely. 

I'd also say based on  previous challenges that the mechanism that keeps it fixed should also be upfront (when the new item is added) rather than relying on event based removal, even if event based removal is still present (and keeps the list small in normal operation).

Reply:
@ctsims Would not the [check](https://github.com/dimagi/commcare-android/blob/updateRequestHeaders/app/src/org/commcare/network/RequestStats.kt#L25) for the first attempt here automatically gurantees that there is only ever one entry in preference per Install workflow restricting the max number of these prefs to the # of enum entries [here](https://github.com/dimagi/commcare-core/blob/updateRequestHeaders/src/main/java/org/commcare/resources/model/InstallRequestSource.java#L6).

Reply:
I see, I totally misread this code. I thought it was one bucket per Install Attempt per Source (max of 6 buckets per Install attempt), not just one bucket per source for a max of 6 buckets total. 

Thanks for clarifying, you can close this out in that case.